openupgrade_scripts/scripts/hr/16.0.1.1/upgrade_analysis_work.txt

```diff
@@ -0,0 +1,59 @@
+---Models in module 'hr'---
+model hr.contract.type (moved from hr_contract)
+# DONE: moved from hr_contract to hr
+---Fields in module 'hr'---
+hr           / hr.contract.type         / __last_update (datetime)      : previously in module hr_contract
+hr           / hr.contract.type         / _order                        : previously in module hr_contract
+hr           / hr.contract.type         / display_name (char)           : previously in module hr_contract
+hr           / hr.contract.type         / name (char)                   : previously in module hr_contract
+# DONE: moved from hr_contract to hr
+hr           / hr.contract.type         / sequence (integer)            : NEW
+# NOTHING TO DO
+hr           / hr.department            / master_department_id (many2one): NEW relation: hr.department, isfunction: function, stored
+hr           / hr.department            / parent_path (char)            : NEW
+hr           / hr.department            / plan_ids (one2many)           : NEW relation: hr.plan
+# NOTHING TO DO
+hr           / hr.employee              / mobile_phone (char)           : now a function
+hr           / hr.employee              / work_contact_id (many2one)    : NEW relation: res.partner
```

Comment:
I think we should call `_inverse_work_ontact_details` to create the `work_contact_id` records.
See https://github.com/odoo/odoo/blob/5ac58ebf983c1c02019253c0430fde3502e13e7c/addons/hr/models/hr_employee_base.py#L169

Reply:
Well spotted. Indeed, we need to compute the now `work_contact_id` and filling it with information that already exist in the database.

Reply:
I think we shouldn't create work_contact_id for old data, because this field not requires to avoid creating so much junk contact. Indeed, employee may already had contact by create manually

Reply:
@nguyenvietlam0640 But it seams to me that it will be created anyway when the ORM will call the `inverse_work_contact_details()` to fill field like `work_mail` and `mobile_phone`. Isn't it ?

Reply:
I just tested it. And the `inverse_work_contact_details()` method **wasn't called** by ORM when migrate till I had changed the `mobile_phone` or `work_email` of employee and it was called to create a new contact. In addition,  The way you created `work_contact_id` by SQL, the field `active`  of `res_partner` will auto set to `false`  which will set to `true` when handle by ORM.

Reply:
@nguyenvietlam0640 You'r right. I tested also to be sure that previous data for `work_email` and `mobile_phone` was not lost when performing migration.

Thanks a lot for your tests and pointing it out that my function for creating contact was not  correct and not useful. :)

I will drop it.

Reply:
@marielejeune Do you still think it's a good idea to run the `_inverse_work_contact_details()` during the migration process ?

Reply:
Not sure anymore. Maybe let's wait for one more advice?

Reply:
@OCA/openupgrade-maintainers Do you have an advice on this topic ?

To summary:

`hr.employee` are now linked to a `res.partner` to store work information of the employee. Migration works well without doing anything. Meaning that the `res.partner` are not created during migration, but each time you modify an employee work information (email, phone, etc), a compute function is triggered and create a new partner with these informations.

Should we leave the migration like this or should we create all the new `res.partner` during the migration process ?

@nguyenvietlam0640 is in favor of doing nothing because it will create a lot of "junk" contact, and maybe these contact already exists but are not linked to the employee in the database.

@marielejeune and myself were in favor of creating all the partner during the migration process, because at the end it will be created.

I think that the three options are:

1. Do nothing, partner will be created on the fly or outside of the migration script.
2. Run the compute function `_inverse_work_contact_detail()` on each `hr.employee` to create all the partner during the migration process.
3. Create all the partner via SQL requests instead of by running the compute function during the migration process.

Any advice to decide are welcome. :)

Reply:
Doing nothing seems like the worst option to me.

- It is not transparent to a user of the software that they trigger a partner creation by changing the phone or e-mail of an employee. That's _terrible_ UX.
  - If they want to do it, they won't intuitively know how.
  - If they don't want to do it, they may never find out why they're slowly getting these additional partner records.
    - And if they consider these records 'junk', then it's much preferable to be able to deal with all junk records _at once_ instead of slowly having to deal with them over time.
- The module kind of _assumes_ that a partner exists, even if it's not strictly mandatory. It's best to go along with those assumptions.
- In the future, the partner field may become required/mandatory. We'd end up having to deal with corner cases anyway, but data integrity/quality-wise, it seems better to me to bite the bullet earlier on as many records as possible.
- An Odoo migration is the best time for data clean-up. Data records will change; dealing with the new 'junk' partner records is just one more thing on the todo list during the migration process.

In any case, this side effect of migration should be clearly documented.

Writing an SQL query is probably preferable. I may be getting this wrong, but it is my reading that using `_inverse_work_contact_detail()` would practically work, but might break if the user has another module installed that implements that method.

Reply:
I discussed with @legalsylvain and we test the default behaviour in 16.0 regarding employee and `work_contact_id`.

If we create a fresh new contact in 16.0, without setting `mobile_phone` and `work_email` (in fact, setting any field except the name of the employee). The `_inverse_work_contact_details()` is triggered and a new partner is created with empty value.

So it seams clear that in 16.0 an employee must be related to a partner for the field `work_contact_id`.

As the philosophy of OpenUpgrade is that, when migrating from previous version you should end with data that are the same as if you started your project in the target version, the partner related to employee must be created during migration process.

Reply:
Hi @remytms, from my point of view the employees are most of the time configured as follow (till Odoo v15)  : 
1. employee is linked to a user (field user_id which then points to a partner_id which is most of the time configured with parent_id = Your company (Odoo company_id.partner_id))
2. employee is linked to a private address (field address_home_id)

Therefore most of the time your new field work_contact_id already exists and I would try to assign in the following order :

1. employee user_id.partner_id (partner linked to user linked to employee)
2. search for partner with same work_email, type="contact" and parent_id = company_id.partner_id

If none of the above match, you may want to create a new contact.

For information, on Odoo v16 runbot, I get the following when looking at employees :

id | name | work_contact_id | address_home_id | user_partner_id
-- | -- | -- | -- | --
#7 | Marc Demo | 7,YourCompany, Marc Demo | 42,Mark Demo | 7,YourCompany, Marc Demo
#1 | Mitchell Admin | 3,YourCompany, Mitchell Admin | 41,Mitchell Admin | 3,YourCompany, Mitchell Admin